### ISSUE TITLE
`utils.map_array`: error for non-writeable input arrays

### ISSUE THREAD

**Issue by m-albert** [scikit-image/scikit-image#6378]
## Description

An error is thrown when passing a non-writeable numpy array to a function such as `skimage.segmentation.relabel_sequential` that calls `utils.map_array`. This seems unexpected, as the input array should not be modified.

## Way to reproduce
```python
import numpy as np
import skimage

l = np.zeros((1,), dtype=np.uint64)
l.setflags(write=False)

skimage.util.map_array(l, np.array([1]), np.array([1]), out=np.zeros((1,)))
# skimage.segmentation.relabel_sequential(l, 50) # api call that fails (unexpectedly for the user?)
```

> ---------------------------------------------------------------------------
> ValueError                                Traceback (most recent call last)
> Input In [59], in <cell line: 2>()
>       1 # from skimage.util._map_array import map_array
> ----> 2 skimage.util.map_array(l, np.array([1]), np.array([1]), out=np.zeros((1,)))
> 
> File ~/miniconda3/envs/test/lib/python3.9/site-packages/skimage/util/_map_array.py:57, in map_array(input_arr, input_vals, output_vals, out)
>      55 input_vals = input_vals.astype(input_arr.dtype, copy=False)
>      56 output_vals = output_vals.astype(out.dtype, copy=False)
> ---> 57 _map_array(input_arr, out_view, input_vals, output_vals)
>      58 return out
> 
> File ~/miniconda3/envs/test/lib/python3.9/site-packages/skimage/util/_remap.pyx:9, in skimage.util._remap._map_array()
> 
> File stringsource:658, in View.MemoryView.memoryview_cwrapper()
> 
> File stringsource:349, in View.MemoryView.memoryview.__cinit__()
> 
> ValueError: buffer source array is read-only

## Version information
```python
# Paste the output of the following python commands
from __future__ import print_function
import sys; print(sys.version)
import platform; print(platform.platform())
import skimage; print(f'scikit-image version: {skimage.__version__}')
import numpy; print(f'numpy version: {numpy.__version__}')
```

```python
# your output here
3.9.12 (main, Apr  5 2022, 06:56:58) 
[GCC 7.5.0]
Linux-5.15.0-27-generic-x86_64-with-glibc2.35
scikit-image version: 0.19.2
numpy version: 1.21.6
```

**Comment by lagru** [comment]
Hey and sorry for the late reply. From what I can gather this happens because it is passed to a memoryview that is declared without `const` in https://github.com/scikit-image/scikit-image/blob/1ede87167068c9e8fc8871baf6a80fcce20eabe9/skimage/util/_remap.pyx#L9

Since Cython 0.28, [read-only memoryviews](https://cython.readthedocs.io/en/latest/src/userguide/memoryviews.html#read-only-views) are supported. However we can't yet use that here because `np_anyint` is a fused type and `const` only supports that from [Cython 3.0.0 alpha 1](https://cython.readthedocs.io/en/latest/src/changes.html#alpha-1-2020-04-12) onwards (added in https://github.com/cython/cython/issues/1772).

So once that version is released and we bump our Cython build dependency to 3.0.0 the fix would be easy. Not sure yet how to deal with that in the meantime. Perhaps we could [declare additional fused types with `const`](https://github.com/cython/cython/issues/1772#issuecomment-371484338) in [used_numerics.pxd](https://github.com/scikit-image/scikit-image/blob/1ede87167068c9e8fc8871baf6a80fcce20eabe9/skimage/_shared/fused_numerics.pxd)? That would be a way to address this now while keeping track of the "proper fix" in TODO.txt.

**Comment by grlee77** [comment]
> Perhaps we could https://github.com/cython/cython/issues/1772#issuecomment-371484338 in [used_numerics.pxd](https://github.com/scikit-image/scikit-image/blob/1ede87167068c9e8fc8871baf6a80fcce20eabe9/skimage/_shared/fused_numerics.pxd)? That would be a way to address this now while keeping track of the "proper fix" in TODO.txt.

Yeah, this seems worth trying. At least, it seems preferable to adding code that would make an unnecessary copy of read-only inputs to satisfy the existing Cython code. Also agree that having things marked as `const` can be helpful to communicate intent to developers.

**Comment by lagru** [comment]
Just took another look: it's not as trivial as I hoped. I couldn't get the solution in https://github.com/cython/cython/issues/1772#issuecomment-371484338 to compile. E.g. I hoped that simply declaring

```python
cdef fused const_anyint:
    const cnp.int8_t
    const cnp.int16_t
    const cnp.int32_t
    const cnp.int64_t
    const cnp.uint8_t
    const cnp.uint16_t
    const cnp.uint32_t
    const cnp.uint64_t

@cython.boundscheck(False)  # Deactivate bounds checking
@cython.wraparound(False)   # Deactivate negative indexing
def _map_array(const_anyint[:] inarr, np_numeric[:] outarr,
               np_anyint[:] inval, np_numeric[:] outval):
    ...
```
would work but that get's me the non-descriptive error

```
  [1/1] Cythonizing /home/lg/Projects/skimage/skimage/util/_remap.pyx

  Error compiling Cython file:
  ------------------------------------------------------------
  ...
      cdef __Pyx_memviewslice memslice
      cdef Py_ssize_t itemsize
      cdef bint dtype_signed
      cdef char kind
      itemsize = -1
      cdef bint ___pyx_const int8_t_is_signed
                            ^
  ------------------------------------------------------------

  (tree fragment):16:27: Syntax error in C variable declaration
```

It may be that we really have to wait for upstream to release for a proper fix. :/

**Comment by m-albert** [comment]
Thanks for investigating the issue @lagru.

> Just took another look: it's not as trivial as I hoped. I couldn't get the solution in https://github.com/cython/cython/issues/1772#issuecomment-371484338 to compile. E.g. I hoped that simply declaring

It seems the workaround only works for native types, see https://github.com/cython/cython/issues/1772#issuecomment-373689862.

Also, I found this issue referencing the same problem and tracking the cython dependency bump as a milestone https://github.com/scikit-image/scikit-image/issues/3492.

**Comment by MiTPenguin** [comment]
Hi the issues seem to have reared its ugly head in another package that depends on skimage  [scverse/spatialdata#706](https://github.com/scverse/spatialdata/issues/706). Was there any update on a workaround?

**Comment by stefanv** [comment]
Looks like it works with Cython 3.0.10: https://github.com/scikit-image/scikit-image/pull/7535

We should apply this across the library.

The only workaround we have for now is to pass in an array with the writable flag set. We promise we don't write to it :)
